### PR TITLE
feat: strcomp

### PR DESCRIPTION
**Pull request by tony-go**
⭐  implement `strcomp`

### CODE REVIEW

File: src/string.c [hunk 1]
```diff
@@ -14,4 +16,23 @@ size_t strlen(const char *s) {
...
 }
 
+int strcmp(const char *s1, const char *s2) {
+  printf("zoo");
+  printf("%c, - %c \n", *s1, *s2);
+
+  char string_one = *s1;
+  char string_two = *s2;
+
+  while (string_one != 0 || string_two != 0) {
```
Comment (capitnflam) [inline src/string.c]:
```suggestion
  while (string_one != 0 && string_two != 0) {
```
If you use `||` you will continue to loop after you have reached the end of the shortest string.

File: src/string.c [hunk 2]
```diff
@@ -14,4 +16,23 @@ size_t strlen(const char *s) {
...
+    printf("%c, - %c \n", string_one, string_two);
+    if (string_one == string_two) {
+      string_one++, string_two++;
+    } else {
+      return 0; // false
+    }
+  }
+
+  return 1; // true
```
Comment (capitnflam) [inline src/string.c]: [strcmp](https://linux.die.net/man/3/strcmp)
> The strcmp() and strncmp() functions return an integer less than, equal to, or greater than zero if s1 (or the first n bytes thereof) is found, respectively, to be less than, to match, or be greater than s2.

File: src/string.c [hunk 3]
```diff
@@ -14,4 +16,23 @@ size_t strlen(const char *s) {
...
+  printf("%c, - %c \n", *s1, *s2);
+
+  char string_one = *s1;
+  char string_two = *s2;
+
+  while (string_one != 0 || string_two != 0) {
+    printf("%c, - %c \n", string_one, string_two);
+    if (string_one == string_two) {
+      string_one++, string_two++;
+    } else {
```
Comment (capitnflam) [inline src/string.c]: you are incrementing the character here not the pointer.

**Review by capitnflam — CHANGES_REQUESTED** [review]
Good start!

File: src/string_lib.c [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+int str_cmp(const char* s1, const char* s2) {
+  int res = 0;
+
+  while (*s1 != 0 && *s2 != 0) {
+    char buffer_1 = *s1;
+    char buffer_2 = *s2;
+    // printf("%c, %c \n", buffer_1, buffer_2);
+
+    if (buffer_1 != 0 && buffer_2 == 0) {
```
Comment (capitnflam) [inline src/string_lib.c]: This will never happen as the loop would have stopped before that.

File: src/string_lib.c [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      printf("here a");
+      res++;
+      while (*(++s1) != 0) {
+        res++;
+      }
+      break;
+    }
+
+    if (buffer_1 == 0 && buffer_2 != 0) {
```
Comment (capitnflam) [inline src/string_lib.c]: Same as above.

**Review by capitnflam — CHANGES_REQUESTED** [review]
@tony-go The basic algo for string comparison is to loop until the string are different (and the end is not reached) and after this just compare the 2 characters where the looped stopped.
This will tell you all you need to know.

**Comment by tony-go** [comment]
Seems to be good! @capitnflam ^^

I just extract the two sub loops outside of the first one 🐐.

File: src/string_lib.c [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  return length;
+}
+
+int str_cmp(const char *s1, const char *s2)
+{
+  int res = 0;
+
+  while (*s1 != 0 && *s2 != 0)
```
Comment (capitnflam) [inline src/string_lib.c:22]:
```suggestion
  while (*s1 != 0 && *s2 != 0 && *s1 == *s2)
```
You need to compare the strings as you go along!

File: src/string_lib.c [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  if (*s1 == 0 && *s2 != 0)
+  {
+    res--;
+    while (*(++s2) != 0)
+    {
+      res--;
+    }
+  }
```
Comment (capitnflam) [inline src/string_lib.c]: This is not necessary.
Once you have done the first loop there is no need to loop anymore.
You just need to compare `*s1` and `*s2`.

```suggestion
  if (*s1 == *s2) {
    return 0;
  }
  if (*s1 < *s2) {
    return -1;
  }
  return 1; // No need for a comparison here as it is the last possible case!
```

File: src/string_lib.c [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+  return res;
```
Comment (capitnflam) [inline src/string_lib.c]: And of course this is not necessary anymore if you take the above code!

File: src/main.c [hunk 9]
```diff
@@ -1,30 +1,61 @@
...
-  printf("- assert empty string %s\n", assert_int(strlen(""), 0));
-  printf("- assert null string %s\n", assert_int(strlen(NULL), 2));
+  printf("-> str_cmp:\n");
+  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yo"), 0));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yoo", "yo"), 1));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yooo", "yo"), 2));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yoooo", "yo"), 3));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yoo"), -1));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yooo"), -2));
+}
```
Comment (capitnflam) [inline src/main.c]:
```suggestion
  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yo"), 0));
  printf("- assert equal string %s\n", assert_int(str_cmp("yoo", "yo"), 1));
  printf("- assert equal string %s\n", assert_int(str_cmp("yooo", "yo"), 1));
  printf("- assert equal string %s\n", assert_int(str_cmp("yoooo", "yo"), 1));
  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yoo"), -1));
  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yooo"), -1));
  printf("- assert equal string %s\n", assert_int(str_cmp("abc", "bcd"), -1));
  printf("- assert equal string %s\n", assert_int(str_cmp("bcd", "abc"), 1));
}
```
The last 2 are cases that would be `0` with your current code.

**Review by capitnflam — CHANGES_REQUESTED** [review]

File: src/string_lib.c (hunk 7 above)
Comment (tony-go) [reply src/string_lib.c]: "less than, equal to, or greater than zero" confuses me a lot.

File: src/main.c [hunk 10]
```diff
@@ -1,30 +1,61 @@
...
-  printf("- assert base length %s\n", assert_int(strlen("yo"), 2));
-  printf("- assert empty string %s\n", assert_int(strlen(""), 0));
-  printf("- assert null string %s\n", assert_int(strlen(NULL), 2));
+  printf("-> str_cmp:\n");
+  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yo"), 0));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yoo", "yo"), 1));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yooo", "yo"), 1));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yoooo", "yo"), 1));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yoo"), -1));
+  printf("- assert equal string %s\n", assert_int(str_cmp("yo", "yooo"), -1));
```
Comment (capitnflam) [inline src/main.c:24]: You should really add a few more tests here, as you are only testing for string length at the moment.

File: src/string_lib.c (hunk 6 above)
Comment (capitnflam) [reply src/string_lib.c:22]: @tony-go You are still not comparing the strings here!

**Review by capitnflam — CHANGES_REQUESTED** [review]
@tony-go There are still some issues with this code.
Please add a test that compares `"abc"` & `"bcd"` you should be able to understand your problem.

File: src/main.c (hunk 10 above)
Comment (tony-go) [reply src/main.c:24]: 😵‍💫

**Review by capitnflam — APPROVED** [review]

**Pull request merged**